### DISCUSSION TITLE
Touch GPIO, looking for examples, clarifications

### DISCUSSION THREAD

**Discussion by SimonFili** [arendst/Tasmota#16596] [Q&A]
Hi there,

Reading this page: https://tasmota.github.io/docs/TouchPin/
I cannot make sense of how to use it.

1) The command explanations between TouchCal and TouchThres are the same, does not seem right
 - I cannot seem to find how to set a threshold, searched here, the net, YouTube, did not find any example 

2) What I tried
 - I've setup GPIO 8 on an ESP32 S3 board to button_tc
 - Connected the pin to metal
 - sent this command "Touchcal 1"
 
01:45:47.871 CMD: touchcal 1
01:45:47.872 Button Touchvalue Hits,
01:45:47.875 MQT: stat/ESP32S3_T1/RESULT = {"TouchCal": 2}
01:45:47.886 PLOT: 1, 56843, 0,

If I touch I get:
01:46:47.305 PLOT: 1, 192005, 0,

So I can see a difference when I touch, going from around 56k to 192k.

I cannot understand how to set a threshold from reading the doc:
> TouchThres x | x=button 1 .. 4. This plots the sensor values to the console, to get information regarding the setting of the 2 following commands

Which is the same text as the command touchcal?

So, the questions are:
1) Is there a working example of how to use this?
2) Can I, after I discover how to set a threshold, get an MQTT message when the pin is touched?
3) Also, the doc talks about a max of 4 on the button_c, but it's 8 on Tasmota 12.1.1
4) What is the meaning of this "RESULT = {"TouchCal": 2}" ?
5) The doc say:
> The continuous pin reading gives a unitless value, that will decrease when the pin (or connected cable) is touched.

But I see the reverse, it increases when touched ?

Thanks!

**Comment by nccchirag** [comment]
you will need to modify the source and compile a custom build. As mentioned in my [comment](https://github.com/arendst/Tasmota/discussions/16518#discussioncomment-3652175) 
the threshold is currently capped at 255 uint8_t which is the main cause of non-functional feature

So, whatever we try to set using console, the MAX number saved is 255 and as you observed the voltage values returned are much larger. 

Basic working can be achieved by modifying 2 files

1. support_command.ino -- comment the lines below
![image](https://user-images.githubusercontent.com/7569367/191507440-f832e17a-f016-4ac0-a9df-f2521605a131.png)

2.. support_button _v3.ino -- change the datatype to uint32_t
![image](https://user-images.githubusercontent.com/7569367/191507736-b34ae7e4-d01b-4516-a8a6-af1e62d9860a.png)

3. Adding log info provide more insight as to where the program hits
![image](https://user-images.githubusercontent.com/7569367/191511717-2c94248f-424a-49f4-b3cc-b06497567e06.png)

**My Test Observations**

1. Tested on ESP32 & ESP32-S3 with no improvements in performance 
2. Internally Adafruit_seesaw::touchRead is used to read the analog value of a capacitive touch-enabled pin which may require improvements too
3. multi-function detection does not work - single tap, double tap & hold (the logic is built for tactile buttons & touch)
5. there are false triggers - mostly HOLD action
6. Basic IDF 4.4.2 [sample](https://github.com/espressif/esp-idf/tree/master/examples/peripherals/touch_sensor) works really well on ESP32-S3
7. Serial Plotter python script works well and help us identify the touch pin / channel
8. TouchThres is set for all pins & NOT per channel
9. TouchCal helps to find the touch value based on which you can set the Threshold

**Conclusion**
Lot of code change is required for the touch feature to work reliably. You may choose that path or use IDF for your TOUCH applications

**Reply by SimonFili** [reply]
@nccchirag Wow, thanks for all the details. I was planning to use touch for detecting water leaking on the floor.  I will not go the route of compiling a specific firmware, because I need Tasmota to be as close as possible to standard versions.  

The other route I will explore is simply to use an analog input for this.

I'm wondering now if the code change you suggest would be a useful addition to Tasmota?

Thanks again to have taken the time to answer.

**Comment by arendst** [comment]
Try latest dev release (v12.1.1.3) It contains a rewrite of the touch driver and now also supports S2 and S3  with default threshold of 40000 as defined in ``my_user_config.h``:

```
#ifdef SOC_TOUCH_VERSION_1                       // ESP32
  #define ESP32_TOUCH_THRESHOLD   40             // [TouchThres] Below this level a touch is detected
#endif
#ifdef SOC_TOUCH_VERSION_2                       // ESP32-S2 and ESP32-S3
  #define ESP32_TOUCH_THRESHOLD   40000          // [TouchThres] Above this level a touch is detected
#endif
```

The ``TouchThres`` parameter is now also persistent so no need for a rule.

Oh and it now also supports multi-press and hold detection.

**Reply by nccchirag** [reply]
you rock @arendst 👏
Thank you so much for the update :) can't wait to test it out
